pkg/trace/providers.go

```diff
@@ -0,0 +1,68 @@
+package trace
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/getoutreach/gobox/internal/call"
+	"github.com/getoutreach/gobox/pkg/log"
+	"github.com/getoutreach/gobox/pkg/orerr"
+	"github.com/getoutreach/gobox/pkg/statuscodes"
+)
+
+// provider is a generic interface for tracking calls.
+type provider interface {
+	Start(ctx context.Context, info *call.Info) context.Context
+	End(ctx context.Context, info *call.Info)
+}
+
+// manyProviders just wraps multiple providers
+type manyProviders []provider
+
+func (px manyProviders) Start(ctx context.Context, info *call.Info) context.Context {
+	for _, p := range px {
+		ctx = p.Start(ctx, info)
+	}
+	return ctx
+}
+
+func (px manyProviders) End(ctx context.Context, info *call.Info) {
+	for idx := range px {
+		px[len(px)-1-idx].End(ctx, info)
+	}
+}
+
+var providers = manyProviders{&defaultCallProvider{}}
+
+// AddProvider is an experimental interface to add a provider.
+//
+// Note that the provider interface is private as this is still experimental.
+// Use x/tracelog.New for an experimental trace log provider.
+func AddProvider(p provider) {
```

Comment:
Note that while this function is public, there is no real way to use this with anything other than pkg/x/tracelog.  Also, added this to avoid using a special trace.yaml config (based on feedback on the proposal).